Objects/typeobject.c

```diff
@@ -3435,25 +3436,43 @@ PyType_FromModuleAndSpec(PyObject *module, PyType_Spec *spec, PyObject *bases)
         goto fail;
     }
 
+    type = &res->ht_type;
+    /* The flags must be initialized early, before the GC traverses us */
+    type->tp_flags = spec->flags | Py_TPFLAGS_HEAPTYPE;
+
     /* Set the type name and qualname */
     const char *s = strrchr(spec->name, '.');
-    if (s == NULL)
+    if (s == NULL) {
         s = spec->name;
-    else
+    }
+    else {
         s++;
+    }
 
-    type = &res->ht_type;
-    /* The flags must be initialized early, before the GC traverses us */
-    type->tp_flags = spec->flags | Py_TPFLAGS_HEAPTYPE;
     res->ht_name = PyUnicode_FromString(s);
-    if (!res->ht_name)
+    if (!res->ht_name) {
+        goto fail;
+    }
+    res->ht_qualname = Py_NewRef(res->ht_name);
+
+    /* Copy spec->name to a buffer we own.
+    *
+    * Unfortunately, we can't use tp_name directly (with some
+    * flag saying that it should be deallocated with the type),
+    * because tp_name is public API and may be set independently
+    * of any such flag.
+    * So, we use a separate buffer, _ht_tpname, that's always
+    * deallocated withthe type (if it's non-NULL).
+    */
+    Py_ssize_t name_buf_len = strlen(spec->name) + 1;
+    res->_ht_tpname = PyMem_Malloc(name_buf_len);
+    if (res->_ht_tpname == NULL) {
         goto fail;
-    res->ht_qualname = res->ht_name;
-    Py_INCREF(res->ht_qualname);
-    type->tp_name = spec->name;
+    }
+    memcpy(res->_ht_tpname, spec->name, name_buf_len);
+    type->tp_name = res->_ht_tpname;
```

Comment:
Feel free to ignore this one, but you can fold these and save one line :)

```suggestion
    type->tp_name = memcpy(res->_ht_tpname, spec->name, name_buf_len);
```